### ISSUE TITLE
Does not work with JSX out of the box in a Next.js project

### ISSUE THREAD

**Issue by dantman** [privatenumber/tsx#500]
### Acknowledgements

- [X] I searched existing issues to avoid duplicates
- [X] I understand this is a place to report a confirmed bug (not seek debugging help)
- [X] I understand this is a collaborative open source project, and relies on community contributions
- [X] I have read and understood the [Contribution guide](https://github.com/privatenumber/tsx/blob/develop/CONTRIBUTING.md)

### Minimal reproduction URL

https://stackblitz.com/edit/stackblitz-starters-pvqxrc?file=tsconfig.json

### Version

v4.7.1

### Node.js version

v18.18.0

### Package manager

npm

### Operating system

Windows

### Problem & expected behavior (under 200 words)

Running `tsx` on code using jsx in a project created by Next.js does not work out of the box. Instead you get the React is not defined error.

This appears to be because Next.js uses `"jsx": "preserve",` and compiles the JSX itself instead of setting `"jsx": "react-jsx",`. So tsx ends up running with the old JSX setting instead of the newer react-jsx output.

![image](https://github.com/privatenumber/tsx/assets/53399/9f385c55-c3e6-455b-a7d0-6de49f7bbb59)

### Contributions

- [ ] I will open a pull request for this issue
- [ ] I will support the work financially (any amount helps)

**Comment by dantman** [comment]
Personally I will workaround the issue by adding a custom tsconfig, but it would be nice if there was an easier way for tsx to know that the newer jsx compiler option should be used instead of the old one.
**/tsconfig.node.json**:
```json
{
	"extends": "./tsconfig.json",
	"compilerOptions": {
		"jsx": "react-jsx"
	}
}
```

**Comment by privatenumber** [comment]
This doesn't sound like a bug (Ive never aimed for Next specific compatibility). Maybe a feature request.

But how is this on tsx to improve as opposed to Next.js?

**Comment by AugustinMauroy** [comment]
On nodejs.org we would like to use `node:test` instead of Jest we use TSX. But we end up in `[Error]: React is not defined`. Do you have a viable solution?

**Comment by privatenumber** [comment]
@AugustinMauroy

I appreciate you reaching out about this. I'm not sure if this question belongs in this thread though.

Can you open a Discussion with the code to reproduce the error? I'd be happy to help there.

**Comment by AugustinMauroy** [comment]
@privatenumber I'm working on a POC repo so I don't have all the nodejs.org codebase.

**Comment by AugustinMauroy** [comment]
question, does the loader take  care about `tsconfig.json` ?

**Comment by AugustinMauroy** [comment]
# solution 

```bash
node --import=tsx --import=./react-loader.mjs myFile.ts
```

`react-loader.mjs`:

```mjs
import * as React from "react";

global.React = React;
```

**Comment by dantman** [comment]
> But how is this on tsx to improve as opposed to Next.js?

Fair, though I don't think Next.js is necessarily in any better of a position to fix this.

1. The Next.js tsconfig is `"jsx": "preserve"`, which TypeScript says means ["Emit .jsx files with the JSX unchanged"](https://www.typescriptlang.org/tsconfig#jsx), so that the JSX can be processed by another compiler or bundler before executing.
2. The "React is not defined" error exists when `"jsx": "react"` is used in settings, which outputs `React.createElement`.
3. While `"jsx": "react-jsx"` indicates "Emit .js files with the JSX changed to _jsx calls" which is automatically imported.

So I agree that Next.js is not giving an indication on how the JSX should be treated. Ideally, since Next.js offers stuff like `next/jest` to use the Next.js compiler with Jest it would have something equivalent to the `tsx`/`ts-node` commands to run scripts.

However at the same time the `tsx` command is reading `"jsx": "preserve"` from tsconfig which says "don't transform JSX" but interpreting it as `"jsx": "react"` and transforming JSX to `React.createElement`. The technically correct behaviour is actually a syntax error, so the behaviour of instead executing something that will run the same as if you were passing this TypeScript output into Babel with `jsxRuntime: classic` set on `@babel/plugin-transform-react-jsx` is a `tsx` command behaviour.

There are a couple ways this could be solved, like a CLI arg to define how untransformed `.jsx` output is supposed to execute in Node.js or detecting when tsx should switch from emitting React.createElement to instead emitting _jsx calls. However, thinking about it the best solution might be to just allow tsx to be configured [the way ts-node can be](https://typestrong.org/ts-node/docs/configuration#via-tsconfigjson-recommended).

Basically, the solution would be if the user needs to run `tsx` on code that uses a separate bundler/compiler to transform JSX then the tsconfig can include config telling `tsx` what to do instead.
```json
{
  "compilerOptions": {
    ...
    "jsx": "preserve",
    ...
  },
  "tsx": {
    "compilerOptions": {
      "jsx": "react-jsx"
    }
  }
}
```

Or you could read `ts-node.compilerOptions` if you don't want `tsx` specific config.

**Comment by privatenumber** [comment]
Sorry, but please keep your messages shorter.

I'm not sure if this is a problem on tsx to solve. tsx aims for TypeScript parity and this seems out of scope.

Please let me know if you can find a disparity with the TypeScript compiler in relation to this.
